ownCloud/Gallery/AlternativeCollectionCollectionViewController.swift

```diff
@@ -0,0 +1,157 @@
+	}
+
+	// MARK: - ViewController LifeCycle
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        // Register cell classes
+        self.collectionView!.register(GalleryCollectionViewCell.self, forCellWithReuseIdentifier: reuseIdentifier)
+		self.collectionView.isPagingEnabled = true
+		self.collectionView.scrollToItem(at: IndexPath(row: index, section: 0), at: .centeredHorizontally, animated: false)
+    }
+
+    // MARK: UICollectionViewDataSource
+    override func numberOfSections(in collectionView: UICollectionView) -> Int {
+        return 1
+    }
+
+    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return items.count
+    }
+
+    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: reuseIdentifier, for: indexPath) as? GalleryCollectionViewCell
+
+		let itemToDisplay = items[indexPath.row]
+		let mimeType = itemToDisplay.mimeType!
+
+		let viewController = self.selectDisplayViewControllerBasedOn(mimeType: mimeType)
+		let shouldDownload = viewController is (DisplayViewController & DisplayExtension) ? true : false
+
+		var configuration: DisplayViewConfiguration
+		if !shouldDownload {
+			configuration = DisplayViewConfiguration(item: itemToDisplay, core: core, state: .notSupportedMimeType)
+		} else {
+			configuration = DisplayViewConfiguration(item: itemToDisplay, core: core, state: .hasNetworkConnection)
+		}
+
+		if cell?.item == nil {
+			cell?.item = itemToDisplay
+			viewController.configure(configuration)
+			self.addChild(viewController)
+			cell?.contentView.addSubview(viewController.view)
+			viewController.didMove(toParent: self)
+			viewController.present(item: itemToDisplay)
+		} else {
+			if cell?.item != itemToDisplay {
+				cell?.item = itemToDisplay
+				viewController.configure(configuration)
+				self.addChild(viewController)
+				cell?.contentView.addSubview(viewController.view)
+				viewController.didMove(toParent: self)
+				viewController.present(item: itemToDisplay)
+			}
+		}
+
+        return cell!
+    }
+
+	// MARK: - Extension selection
+	private func selectDisplayViewControllerBasedOn(mimeType: String) -> (DisplayViewController) {
+
+		let locationIdentifier = OCExtensionLocationIdentifier(rawValue: mimeType)
+		let location: OCExtensionLocation = OCExtensionLocation(ofType: .viewer, identifier: locationIdentifier)
+		let context = OCExtensionContext(location: location, requirements: nil, preferences: nil)
+
+		var extensions: [OCExtensionMatch]?
+
+		do {
+			try extensions = OCExtensionManager.shared.provideExtensions(for: context)
+		} catch {
+			return DisplayViewController()
+		}
+
+		guard let matchedExtensions = extensions else {
+			return DisplayViewController()
+		}
+
+		let preferedExtension: OCExtension = matchedExtensions[0].extension
```

Comment:
@pablocarmu can `matchedExtensions` be an empty array?

Reply:
IIRC if there are no extensions, the SDK returns nil.
In any case I'll code the check for the count to be more futureproof